llvm/lib/Transforms/Scalar/SROA.cpp

```diff
@@ -5454,6 +5475,79 @@ void SROA::clobberUse(Use &U) {
     }
 }
 
+/// A basic LoadAndStorePromoter that does not remove store nodes.
+class BasicLoadAndStorePromoter : public LoadAndStorePromoter {
+public:
+  BasicLoadAndStorePromoter(ArrayRef<const Instruction *> Insts, SSAUpdater &S,
+                            Type *ZeroType)
+      : LoadAndStorePromoter(Insts, S), ZeroType(ZeroType) {}
+  bool shouldDelete(Instruction *I) const override {
+    return !isa<StoreInst>(I) && !isa<AllocaInst>(I);
+  }
+
+  Value *getValueToUseForAlloca(Instruction *I) const override {
+    return Constant::getNullValue(ZeroType);
```

Comment:
```suggestion
    return UndefValue::get(ZeroType);
```

Reply:
I thought we were trying to remove uses of undef?

Reply:
Generally yes. Uninitialized memory is the exception to this currently.